### ISSUE TITLE
My Fixture is initialized more than once even though I use CollectionFixture

### ISSUE THREAD

**Issue by csimone86** [xunit/xunit#2565]
Hi everyone,
I am writing some test classes in `.net6` using `xUnit 2.4.1` in VS2022.

I have two assemblies:
- The first one contains the Fixture;
- The second one the tests + the collection.

What I expect is that I run into the constructor of the Fixture class just once. But what happen is to run into it different times.

Here all my codes:

First assembly, `the Fixture` class:

    public class MyDbContextFixture 
    {
        private Guid x = Guid.Empty;

        public MyDbContextFixture()
        {
            x = Guid.NewGuid();

            System.Diagnostics.Debug.WriteLine($"Ctor Fixture. Guid {x}. ");
        }

        public void Cleanup()
        {
            System.Diagnostics.Debug.WriteLine($"Dispose Fixture. Guid {x}.");
        }
    }

The second assembly, `the Collection` class:

    [CollectionDefinition("TestsCollectionsAdmin")]
    public class AdminTestsCollection : ICollectionFixture<MyDbContextFixture>
    {
    }

The second assembly, `the First Tests` class:

    [Collection("TestsCollectionsAdmin")]
    public class TestsA : IDisposable
    {
        private MyDbContextFixture _context { get; }

        public TestsA(MyDbContextFixture context)
        {
            _context = context;
            System.Diagnostics.Debug.WriteLine("Ctor Test class A");
        }

        [Fact]
        public void Test1()
        {
            System.Diagnostics.Debug.WriteLine("Test class A -> Test 1");
        }

        [Fact]
        public void Test2()
        {
            System.Diagnostics.Debug.WriteLine("Test class A -> Test 2");
        }

        public void Dispose()
        {
            System.Diagnostics.Debug.WriteLine("Dispose Test class A");
            _context.Cleanup();
        }
    }

The second assembly, `the Second Tests` class:

    [Collection("TestsCollectionsAdmin")]
    public class TestsB : IDisposable
    {
        private MyDbContextFixture _context { get; }

        public TestsB(MyDbContextFixture context)
        {
            _context = context;
            System.Diagnostics.Debug.WriteLine("Ctor Test class B");
        }

        [Fact]
        public void Test1()
        {
            System.Diagnostics.Debug.WriteLine("Test class B -> Test 1");
        }

        [Fact]
        public void Test2()
        {
            System.Diagnostics.Debug.WriteLine("Test class B -> Test 2");
        }

        [Fact]
        public void Test3()
        {
            System.Diagnostics.Debug.WriteLine("Test class B -> Test 3");
        }

        public void Dispose()
        {
            System.Diagnostics.Debug.WriteLine("Dispose Test class B");
            _context.Cleanup();
        }
    }

The result is:

> Ctor Fixture. Guid 2049ef03-3104-4ba1-835f-25e288041385. 
> Ctor Fixture. Guid e400b28f-50ac-4da5-a007-40bb0f608bd9. 
> Ctor Fixture. Guid e11651c5-ac58-4371-b743-3663c072b157. 
> Ctor Fixture. Guid 29e63556-fcba-444f-b20f-ef66eff87925. 
> Ctor Test class A
> Ctor Test class B
> Ctor Test class B
> Ctor Test class B
> Test class B -> Test 2
> Test class B -> Test 1
> Test class A -> Test 1
> Test class B -> Test 3
> Dispose Test class B
> Dispose Test class A
> Dispose Fixture. Guid 2049ef03-3104-4ba1-835f-25e288041385.
> Dispose Fixture. Guid e400b28f-50ac-4da5-a007-40bb0f608bd9.
> Dispose Test class B
> Dispose Test class B
> Dispose Fixture. Guid e11651c5-ac58-4371-b743-3663c072b157.
> Ctor Test class A
> Dispose Fixture. Guid 29e63556-fcba-444f-b20f-ef66eff87925.
> Test class A -> Test 2
> Dispose Test class A
> Dispose Fixture. Guid e400b28f-50ac-4da5-a007-40bb0f608bd9.

What I have found strange is that I go into the Fixture constructor 4 times! Not 1. Not 5 as my tests! 
It is clear and it is correct I go into the cleanup 5 times. Because the dispose is called after each single test. But I do not understand why I go inside the constructor of the fixture 4 times. 

What am I doing wrong?

Thank you

EDIT: 
The problem seems to be that my Fixture is inside another assembly.. If I move everything inside the same assebly, everything works:

> Ctor Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.
>  
> Ctor Test class A
> Test class A -> Test 1
> Dispose Test class A
> Dispose Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.
> 
> Ctor Test class A
> Test class A -> Test 2
> Dispose Test class A
> Dispose Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.
> 
> Ctor Test class B
> Test class B -> Test 2
> Dispose Test class B
> Dispose Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.
> 
> Ctor Test class B
> Test class B -> Test 3
> Dispose Test class B
> Dispose Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.
> 
> Ctor Test class B
> Test class B -> Test 1
> Dispose Test class B
> Dispose Fixture. Guid 4491235a-fcf1-46e8-b6fe-618f433b4134.

Then, I suppose that @bradwilson says [here](https://github.com/xunit/xunit/issues/374#issuecomment-90281871) is partially correct. He said:

> This is supported:
> 
> - Assembly A: Fixture class
> - Assembly B: Collection definition class, unit test class

Well, I don't if we can say that is supported. The behaviour I detected is really different than a Fixture should work.

**Comment by bradwilson** [comment]
I am unable to reproduce what you're seeing.

I simplified the example even further (and tested in both .NET Framework and .NET Core), and made it so the test framework can tell you if it's failing or not (rather than trying to capture and inspect debug output by hand).

*Fixture\Fixture.csproj*

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>

</Project>
```

*Fixture\MyDbContextFixture.cs*

```csharp
public class MyDbContextFixture
{
    public static int InitCount = 0;

    public MyDbContextFixture()
    {
        ++InitCount;
    }
}
```

*Tests\Tests.csproj*

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net48;net6.0</TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\Fixture\Fixture.csproj" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.2.0" />
    <PackageReference Include="xunit" Version="2.4.2" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

*Tests\Collection.cs*

```csharp
using Xunit;

[CollectionDefinition("TestsCollectionsAdmin")]
public class AdminTestsCollection : ICollectionFixture<MyDbContextFixture>
{
}
```

*Tests\TestsA.cs*

```csharp
using Xunit;

[Collection("TestsCollectionsAdmin")]
public class TestsA
{
    public TestsA(MyDbContextFixture _)
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }

    [Fact]
    public void Test1()
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }

    [Fact]
    public void Test2()
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }
}
```

*Tests\TestsB.cs*

```csharp
using Xunit;

[Collection("TestsCollectionsAdmin")]
public class TestsB
{
    public TestsB(MyDbContextFixture _)
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }

    [Fact]
    public void Test1()
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }

    [Fact]
    public void Test2()
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }

    [Fact]
    public void Test3()
    {
        Assert.Equal(1, MyDbContextFixture.InitCount);
    }
}
```

All five tests pass in both target frameworks:

```
 pwsh> dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  Fixture -> C:\Dev\1\Fixture\bin\Debug\netstandard2.0\Fixture.dll
  Tests -> C:\Dev\1\Tests\bin\Debug\net48\Tests.dll
  Tests -> C:\Dev\1\Tests\bin\Debug\net6.0\Tests.dll
Test run for C:\Dev\1\Tests\bin\Debug\net48\Tests.dll (.NETFramework,Version=v4.8)
Microsoft (R) Test Execution Command Line Tool Version 17.2.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     5, Skipped:     0, Total:     5, Duration: 6 ms - Tests.dll (net48)
Test run for C:\Dev\1\Tests\bin\Debug\net6.0\Tests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.2.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     5, Skipped:     0, Total:     5, Duration: 2 ms - Tests.dll (net6.0)
```

If you have a repro, please attach the full repro project (ideally as a GitHub project) along with complete step-by-step instructions on how to reproduce it, because given the information you provided, I am unable to reproduce your issue.

**Comment by bradwilson** [comment]
@csimone86 @Chaosflo Without a full repro project, there's nothing I can do here, since my attempt to create a repro based on the textual description has failed.

**Comment by bradwilson** [comment]
Closing for age. Please open a new issue with a full repro project. Thanks!

**Comment by timabell** [comment]
Belatedly, but this kind of stuff is of interest to me...

I don't think this is thread-safe, and the tests run in parallel so that might be a problem:

```csharp
    public static int InitCount = 0;
```
I think it needs concurrency-proofing for that example to be truly valid. [My example uses a concurrent bag to be safe](https://github.com/timabell/xunit-collections-example/blob/6aa1c3e5fff6cba2d6405f78c8566e8c28ee95c1/CollectionFixture/CollectionFixtureExampleTests.cs#L15).

If anyone might find it useful I have some minimal examples of various shared fixture things here https://github.com/timabell/xunit-collections-example

**Comment by bradwilson** [comment]
@timabell I agree that, if concurrency were expected, then something like `Interlocked` makes the most sense. If no concurrency were expected, then the simple `++` should be fine.
